### PR TITLE
Forward compatible updates for TS 2.4

### PR DESCRIPTION
**Pull request by kitsonk**
**Type:** bug / feature

The following has been addressed in the PR:

* [x] There is a related issue
* [X] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [X] Unit or Functional tests are included in the PR

**Description:**

This PR contains forward compatible updates for TypeScript 2.4

Refs: dojo/meta#189

### CODE REVIEW

**Comment by kitsonk** [comment]
Hmmm... I think I have configured appveyor to ignore branches that don't have a configuration file for appveyor and should stop these from being marked failed.

File: src/async/ExtensiblePromise.ts [hunk 1]
```diff
@@ -140,16 +203,14 @@ export default class ExtensiblePromise<T> {
 	 *
 	 * @returns {ExtensiblePromise}
 	 */
-	then<U, V>(onFulfilled: ((value: T) => (U | Thenable<U> | undefined)) | undefined, onRejected: (reason: Error) => (V | Thenable<V>)): ExtensiblePromise<U | V>;
-	then<U>(onFulfilled?: ((value: T) => (U | Thenable<U> | undefined)) | undefined, onRejected?: (reason: Error) => void): ExtensiblePromise<U>;
-	then<U>(onFulfilled?: ((value: T) => (U | Thenable<U> | undefined)) | undefined, onRejected?: (reason: Error) => (U | Thenable<U>)): ExtensiblePromise<U> {
-		const e: Executor<U> = (resolve, reject) => {
-			function handler(rejected: boolean, valueOrError: T | U | Error) {
-				const callback: ((value: T | U | Error) => (U | Thenable<U> | void)) | undefined = rejected ? onRejected : onFulfilled;
+	then<TResult1 = T, TResult2 = never>(onFulfilled?: ((value: T) => TResult1 | PromiseLike<TResult1>) | undefined | null, onRejected?: ((reason: any) => TResult2 | PromiseLike<TResult2> | void) | undefined | null): ExtensiblePromise<TResult1 | TResult2> {
```
Comment (agubler) [inline src/async/ExtensiblePromise.ts]: This is quite a long line now.

File: src/async/ExtensiblePromise.ts [hunk 2]
```diff
@@ -140,16 +203,14 @@ export default class ExtensiblePromise<T> {
...
+		const e: Executor<TResult1> = (resolve, reject) => {
```
Comment (agubler) [inline src/async/ExtensiblePromise.ts]: I know it was `e` before but we could probably give it a better name.

**Review by agubler — APPROVED** [review]
Just a couple of comments, some super nice documentation additions.

**Pull request merged**